scipy/interpolate/_bspl.pyx

```diff
@@ -434,18 +435,19 @@ def _make_design_matrix(const double[::1] x,
 
     Returns
     -------
-    data, (row_idx, col_idx)
-        Constructor parameters for a CSR matrix: in each row all the basis
+    design matrix
+        Design matrix as CSR matrix: in each row all the basis
         elements are evaluated at the certain point (first row - x[0],
         ..., last row - x[-1]).
     """
     cdef:
-        cnp.npy_intp i, ind
+        cnp.npy_intp i, j, ind
         cnp.npy_intp n = x.shape[0]
         double[::1] work = np.empty(2*k+2, dtype=float)
         double[::1] data = np.zeros(n * (k + 1), dtype=float)
-        cnp.ndarray[long, ndim=1] row_ind = np.zeros(n * (k + 1), dtype=int)
-        cnp.ndarray[long, ndim=1] col_ind = np.zeros(n * (k + 1), dtype=int)
+        # TODO: When should we switch to int64?
+        cnp.npy_int32[::1] indptr = np.arange(0, (n + 1) * (k + 1), k + 1, dtype=np.int32)
+        cnp.npy_int32[::1] indices = np.zeros(n * (k + 1), dtype=np.int32)
```

Comment:
Hm. If these are typed as int32, we need to check for the unt32 overflow. So, three options: check for 32-bit overflow and bail out fast; always use 64-bit indices; follow what scipy.sparse does and switch the integer dtype depending on the problem size. 
While this is still fresh, always using 64-bit indices does not involve serious backcompat concerns.

Reply:
This is related to https://github.com/scipy/scipy/issues/16774.
We could detect/compute the need of int64 upfront. Will do ...

Reply:
Fixing this makes the code a little more complicated. I went with fused types and preallocation of `indices` and `indptr` in Python, see https://github.com/scipy/scipy/pull/16840/commits/6a8ebf0d204741b3fd0304a03296830bbaa5a2b8.

I don't know, how to test this as it takes a long time to compute such a huge b-spline ouput csr arrays.

Reply:
As info: https://github.com/scikit-learn/scikit-learn/pull/23731 implements a similar logic.